### PR TITLE
Add pallet-signal

### PR DESCRIPTION
**Pull request by vasylenko-yevhen**
Pallet Signal migration.

**DONE:**
- [x] Pallet Signal code migration.
- [x] Mocks, tests.

**TODO:**
- [ ] Benchmarks (blocked by Control pallet)

**Next steps:**

1. Merge Flow, Signal, Control feature branches.
2. Test everything together, update zero-protocol runtime.
3. Flow benchmarks, Signal benchmarks.
4. Control tests and benchmarks.
5. Hardening.

### CODE REVIEW

**Comment by vasylenko-yevhen** [comment]
@2075 there are few moments I would like to clarify before finalizing this piece.  
It would be nice if you could address those moments.

1. **`unlock_balance`**
  Currently this [helper method](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L567) is called in `on_finalize` as well as when calling simple vote extrinsic. As far as I understand from the pallet logic, this should be called only when proposal expires in `on_finalize` and should not be called in extrinsic. Correct me if I am wrong.
2. **Text errors**
  There are few checks in extrinsic, which generate text errors instead of specific error enum value, [example 1](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L255), [example 2](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L243). Were these errors generated like this on purpose or it is fine to replace them with one of Error enum values?
3. **Equal amount of votes**
  It is possible to have equal number of yes/no votes, [example](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L670). What should happen with proposal in that case, should it be considered as accepted or declines?

File: signal/src/voting_structs.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+use frame_support::pallet_prelude::{Encode, Decode};
```
Comment (vovacha) [inline signal/src/voting_structs.rs:1]: I'm not sure if we should keep structs and enums in separate places. For now, it's not critical, but let's discuss it later and use the same approach for all pallets.

**Review by vovacha — APPROVED** [review]
LGTM. We still have a bunch of open questions which are not critical for now. I'd like to merge it, since the build is stable and tested. But during the hardening phase we should finalize our discussion and make appropriate changes to the code base.

File: signal/src/mock.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+};
+use zero_primitives::{Amount, Balance, BlockNumber, CurrencyId, Hash, TokenSymbol};
+
+pub type AccountId = u64;
+
+pub const ACC1: AccountId = 1;
+pub const ACC2: AccountId = 2;
+pub const TREASURY_ACC: AccountId = 3;
+
+pub struct ControlFixture {
```
Comment (vovacha) [inline signal/src/mock.rs:27]: Just a reminder: remove these fixtures after Control and Flow will be merged.

**Comment by 2075** [comment]
>     1. **`unlock_balance`**
>        Currently this [helper method](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L567) is called in `on_finalize` as well as when calling simple vote extrinsic. As far as I understand from the pallet logic, this should be called only when proposal expires in `on_finalize` and should not be called in extrinsic. Correct me if I am wrong.

The funds are collected from the contributors first, therefore they are unreserved and transferred to destination wallet.
after collecting the funds are reserved again and only a fraction unreserved to cover the settlement fees.

> >     2. **Text errors**
>        There are few checks in extrinsic, which generate text errors instead of specific error enum value, [example 1](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L255), [example 2](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L243). Were these errors generated like this on purpose or it is fine to replace them with one of Error enum values?

undecided about these, saw it in other places as well. in general nothing should panic in a live system, therefore i consider those which are more like an internal sanity check and nothing that throws / generates an event, can be text. what do you think?

>     3. **Equal amount of votes**
>        It is possible to have equal number of yes/no votes, [example](https://github.com/playzero/subzero/blob/dev/modules/signal/src/lib.rs#L670). What should happen with proposal in that case, should it be considered as accepted or declines?

based on the current implementation it is a draw and could expire. in that case eventually a restart should be possible?
https://github.com/gamedaoco/gamedao-protocol/blob/e929fc7df24535a67a2c6a8c387a423c6bd684a9/signal/src/lib.rs#L715

we could/should also prepare already for different preset ratios and create enums for there, like more than 1/4, 2/4, 3/4, 4/4 etc.

**Pull request merged**